### PR TITLE
Fused bce logits reduce mean

### PR DESCRIPTION
**Pull request by MARD1NO**

### CODE REVIEW

File: oneflow/user/ops/binary_cross_entropy_with_logits_reduce_mean_op.cpp [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+
+namespace {
+
+Maybe<void> InferTensorDescFn(user_op::InferContext* ctx) {
+  const auto& input_desc = ctx->InputTensorDesc("input", 0);
+  const auto& target_desc = ctx->InputTensorDesc("target", 0);
+  CHECK_EQ_OR_RETURN(input_desc.is_dynamic(), target_desc.is_dynamic());
+  CHECK_EQ_OR_RETURN(input_desc.shape(), target_desc.shape());
+  user_op::TensorDesc* out_desc = ctx->OutputTensorDesc("out", 0);
+  *out_desc->mut_is_dynamic() = input_desc.is_dynamic();
```
Comment (liujuncheng) [inline oneflow/user/ops/binary_cross_entropy_with_logits_reduce_mean_op.cpp:30]:
```suggestion
  *out_desc->mut_is_dynamic() = false;
```
这里输出一定是一个标量了吧

File: oneflow/user/ops/binary_cross_entropy_with_logits_reduce_mean_op.cpp [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+#include "oneflow/core/framework/op_generated.h"
+
+namespace oneflow {
+
+namespace {
+
+Maybe<void> InferTensorDescFn(user_op::InferContext* ctx) {
+  const auto& input_desc = ctx->InputTensorDesc("input", 0);
+  const auto& target_desc = ctx->InputTensorDesc("target", 0);
+  CHECK_EQ_OR_RETURN(input_desc.is_dynamic(), target_desc.is_dynamic());
```
Comment (liujuncheng) [inline oneflow/user/ops/binary_cross_entropy_with_logits_reduce_mean_op.cpp:27]: 这个不需要，可以在kernel里检查形状

File: oneflow/user/ops/binary_cross_entropy_with_logits_reduce_mean_op.cpp [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+  const user_op::TensorDesc& target_desc = ctx->InputTensorDesc("target", 0);
+  CHECK_EQ_OR_RETURN(input_desc.data_type(), target_desc.data_type());
+  *ctx->OutputDType("out", 0) = ctx->InputDType("input", 0);
+
+  return Maybe<void>::Ok();
+}
+Maybe<void> InferGradTensorDescFn(user_op::InferContext* ctx) {
+  const auto& input_desc = ctx->InputTensorDesc("input", 0);
+  const auto& target_desc = ctx->InputTensorDesc("target", 0);
+  CHECK_EQ_OR_RETURN(input_desc.is_dynamic(), target_desc.is_dynamic());
```
Comment (liujuncheng) [inline oneflow/user/ops/binary_cross_entropy_with_logits_reduce_mean_op.cpp:46]: 同理

File: oneflow/user/ops/binary_cross_entropy_with_logits_reduce_mean_op.cpp [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    user_op::InferContext* ctx) {
+  return InferLogicalTensorDesc(ctx);
+}
+
+/* static */ Maybe<void> BinaryCrossEntropyWithLogitsReduceMeanOp::GetSbp(
+    user_op::SbpContext* ctx) {
+  ctx->NewBuilder()
+      .Split(user_op::OpArg("input", 0), 0)
+      .Split(user_op::OpArg("target", 0), 0)
+      .Broadcast(user_op::OpArg("out", 0))
```
Comment (liujuncheng) [inline oneflow/user/ops/binary_cross_entropy_with_logits_reduce_mean_op.cpp:77]: 这个是P而不是B吧

File: oneflow/user/ops/binary_cross_entropy_with_logits_reduce_mean_op.cpp [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+  const auto& target_desc = ctx->InputTensorDesc("target", 0);
+  CHECK_EQ_OR_RETURN(input_desc.is_dynamic(), target_desc.is_dynamic());
+  CHECK_EQ_OR_RETURN(input_desc.shape(), target_desc.shape());
+  user_op::TensorDesc* out_desc = ctx->OutputTensorDesc("out", 0);
+  *out_desc->mut_is_dynamic() = input_desc.is_dynamic();
+  *out_desc->mut_shape() = Shape({});
+  return Maybe<void>::Ok();
+}
+
+Maybe<void> InferDataType_(user_op::InferContext* ctx) {
```
Comment (liujuncheng) [inline oneflow/user/ops/binary_cross_entropy_with_logits_reduce_mean_op.cpp:35]: 这个命名不要以下划线结尾

File: oneflow/user/kernels/binary_cross_entropy_with_logits_reduce_mean.cpp [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+using namespace loss;
+
+template<typename T>
+inline T ComputeMaxVal(const T x) {
+  T y = -x;
+  return y < 0 ? 0 : y;
+}
+
+template<typename T>
+inline T CalSigmoid(const T x) {
```
Comment (liujuncheng) [inline oneflow/user/kernels/binary_cross_entropy_with_logits_reduce_mean.cpp:32]:
```suggestion
inline T Sigmoid(const T x) {
```

File: oneflow/user/kernels/binary_cross_entropy_with_logits_mean_kernel.cu [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+#include <cub/cub.cuh>
+
+namespace oneflow {
+
+namespace user_op {
+
+namespace {
+
+constexpr int32_t kBlockSize = 1024;
+constexpr int32_t kReduceLocalSumBlockSize = 256;
```
Comment (liujuncheng) [inline oneflow/user/kernels/binary_cross_entropy_with_logits_mean_kernel.cu:29]:
```suggestion
constexpr int32_t kReduceLocalSumBlockSize = 256;
```
这个也是越大越好吧

File: oneflow/user/kernels/binary_cross_entropy_with_logits_mean_kernel.cu [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+    auto* out_blob = ctx->Tensor4ArgNameAndIndex("out", 0);
+
+    const int64_t elem_cnt = input_blob->shape_view().elem_cnt();
+
+    const T* input = input_blob->dptr<T>();
+    const T* target = target_blob->dptr<T>();
+    T* out = out_blob->mut_dptr<T>();
+    using ComputeType = typename DefaultComputeType<T>::type;
+
+    if (elem_cnt < kBlockSize) {
```
Comment (liujuncheng) [inline oneflow/user/kernels/binary_cross_entropy_with_logits_mean_kernel.cu:158]: 这个阈值可以调整一下，而且不要用kBlockSize

File: oneflow/user/kernels/binary_cross_entropy_with_logits_mean_kernel.cu [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+      FusedBinaryCrossEntropyWithLogitsReduceMeanKernel<T, T, ComputeType>
+          <<<1, kBlockSize, 0, ctx->stream()->As<ep::CudaStream>()->cuda_stream()>>>(
+              input_blob->dptr<T>(), target_blob->dptr<T>(), out_blob->mut_dptr<T>(),
+              input_blob->shape_view().elem_cnt());
+    } else {
+      auto* tmp_buffer = ctx->Tensor4ArgNameAndIndex("tmp_buffer", 0);
+      const int64_t block_num = (elem_cnt + kBlockSize - 1) / kBlockSize;
+      int launch_block = block_num;
```
Comment (liujuncheng) [inline oneflow/user/kernels/binary_cross_entropy_with_logits_mean_kernel.cu:166]: 这里可以在根据tmp_buffer的大小约束一下

File: oneflow/user/kernels/binary_cross_entropy_with_logits_mean_kernel.cu [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+  }
+  const T* dy_ptr;
+  const T elem_cnt_reciprocal;
+};
+
+template<typename T>
+class BinaryCrossEntropyWithLogitsMeanKernel final : public user_op::OpKernel {
+ public:
+  BinaryCrossEntropyWithLogitsMeanKernel() = default;
+  ~BinaryCrossEntropyWithLogitsMeanKernel() = default;
```
Comment (guo-ran) [inline oneflow/user/kernels/binary_cross_entropy_with_logits_mean_kernel.cu:142]: override = default;

File: oneflow/user/kernels/binary_cross_entropy_with_logits_mean_kernel.cu [hunk 11]
```diff
@@ -0,0 +1,251 @@
...
+}
+
+template<typename T>
+__device__ __forceinline__ T Sigmoid(const T x) {
+  const T half_of_one = static_cast<T>(0.5);
+  return half_of_one * tanh(half_of_one * x) + half_of_one;
+}
+
+template<>
+__device__ __forceinline__ float Sigmoid(const float x) {
```
Comment (guo-ran) [inline oneflow/user/kernels/binary_cross_entropy_with_logits_mean_kernel.cu:103]: float不需要特化吧
```
inline _LIBCUDACXX_INLINE_VISIBILITY float       tanh(float __lcpp_x) _NOEXCEPT       {return ::tanhf(__lcpp_x);}
```

File: oneflow/user/kernels/binary_cross_entropy_with_logits_mean_kernel.cu [hunk 12]
```diff
@@ -0,0 +1,275 @@
...
+               const user_op::OpKernelCache* cache) const override {
+    const auto* input_blob = ctx->Tensor4ArgNameAndIndex("input", 0);
+    const auto* target_blob = ctx->Tensor4ArgNameAndIndex("target", 0);
+    auto* out_blob = ctx->Tensor4ArgNameAndIndex("out", 0);
+
+    int64_t elem_cnt = input_blob->shape_view().elem_cnt();
+
+    if (cache) {
+      // Because `out`'s SBP maybe P or B, we need to use logical_elem_cnt as reduce_mean factor.
+      const auto* bce_cache = dynamic_cast<const BCEWithLogitsReduceMeanKernelCache*>(cache);
```
Comment (MARD1NO) [inline oneflow/user/kernels/binary_cross_entropy_with_logits_mean_kernel.cu:173]: BCE可能最后推出来的是B 或者 P，当为P的时候，应该拿的是总的elem_cnt，因此这里用Logical的elem_cnt

**Review by liujuncheng — APPROVED** [review]

File: oneflow/user/kernels/binary_cross_entropy_with_logits_mean_kernel.cu [hunk 13]
```diff
@@ -0,0 +1,275 @@
...
+
+ private:
+  using user_op::OpKernel::Compute;
+  void Compute(user_op::KernelComputeContext* ctx) const override {
+    const auto* input_blob = ctx->Tensor4ArgNameAndIndex("input", 0);
+    const auto* target_blob = ctx->Tensor4ArgNameAndIndex("target", 0);
+    const auto* dy_blob = ctx->Tensor4ArgNameAndIndex("dy", 0);
+    auto* dx_blob = ctx->Tensor4ArgNameAndIndex("dx", 0);
+
+    const int64_t elem_cnt = input_blob->shape_view().elem_cnt();
```
Comment (guo-ran) [inline oneflow/user/kernels/binary_cross_entropy_with_logits_mean_kernel.cu:222]: 这里的elem_cnt是不是也和前向一样

File: oneflow/user/kernels/binary_cross_entropy_with_logits_mean_kernel.cu [hunk 14]
```diff
@@ -0,0 +1,275 @@
...
+  using user_op::OpKernel::Compute;
+  void Compute(user_op::KernelComputeContext* ctx, user_op::OpKernelState* state,
+               const user_op::OpKernelCache* cache) const override {
+    const auto* input_blob = ctx->Tensor4ArgNameAndIndex("input", 0);
+    const auto* target_blob = ctx->Tensor4ArgNameAndIndex("target", 0);
+    auto* out_blob = ctx->Tensor4ArgNameAndIndex("out", 0);
+
+    int64_t elem_cnt = input_blob->shape_view().elem_cnt();
+
+    if (cache) {
```
Comment (liujuncheng) [inline oneflow/user/kernels/binary_cross_entropy_with_logits_mean_kernel.cu:171]: cache != nullptr

File: oneflow/user/kernels/binary_cross_entropy_with_logits_mean_kernel.cu [hunk 15]
```diff
@@ -0,0 +1,275 @@
...
+      : logical_elem_cnt_(logical_elem_cnt) {}
+  ~BCEWithLogitsReduceMeanKernelCache() override = default;
+
+  int64_t logical_elem_cnt() const { return logical_elem_cnt_; }
+
+ private:
+  const int64_t logical_elem_cnt_;
+};
+
+std::shared_ptr<user_op::OpKernelCache> CreateBCEWithLogitsReduceMeanKernelCache(
```
Comment (liujuncheng) [inline oneflow/user/kernels/binary_cross_entropy_with_logits_mean_kernel.cu:145]: 这个方法在哪里被调用的？

File: oneflow/user/kernels/binary_cross_entropy_with_logits_mean_kernel.cu (hunk 14 above)
Comment (liujuncheng) [reply oneflow/user/kernels/binary_cross_entropy_with_logits_mean_kernel.cu:171]: else是什么行为？

File: oneflow/user/kernels/binary_cross_entropy_with_logits_mean_kernel.cu [hunk 16]
```diff
@@ -0,0 +1,275 @@
...
+      CHECK_NOTNULL(bce_cache);
+      elem_cnt = bce_cache->logical_elem_cnt();
+    }
+
+    const T* input = input_blob->dptr<T>();
+    const T* target = target_blob->dptr<T>();
+    T* out = out_blob->mut_dptr<T>();
+    using ComputeType = typename DefaultComputeType<T>::type;
+
+    if (elem_cnt <= kSingleBlockProcessNumThreshold) {
```
Comment (liujuncheng) [inline oneflow/user/kernels/binary_cross_entropy_with_logits_mean_kernel.cu:183]: 命名上区分一下，一个叫elem_cnt是实际的元素个数
另个是mean的分母

**Review by guo-ran — APPROVED** [review]
